File: python/examples/decrypt_mnemonic.py
import os
from dotenv import load_dotenv
from requests import Session
from pbkdf2 import PBKDF2
from Crypto.Cipher import AES
from Crypto.Hash import keccak
load_dotenv()

api_token = os.getenv("API_TOKEN")
exchange_password = os.getenv("ACCOUNT_PASSWORD")
url = "https://localcoinswap.com/api/v2/profile/encrypted-blob/"

s = Session()
s.headers.update({"Authorization": f"Token {api_token}"})

response = s.get(url).json()

salt_hex = bytes.fromhex(response["salt"])
iterations = int(response["iterations"])
initialization_vector = bytes.fromhex(response["initialization_vector"])
encrypted_account_key = response["encrypted_account_key"]
encrypted_mnemonic = response["encrypted_mnemonic"]
encrypted_mnemonic_iv = bytes.fromhex(response["encrypted_mnemonic_iv"])

# Combine salt and plaintext account password in PBKDF2 function
# to create a secret key.
account_key_secret = PBKDF2(exchange_password, salt_hex, iterations).read(32).hex()
print("accountKeySecret PBKDF2", account_key_secret, len(account_key_secret))

# Use initialization vector and secret key to decrypt the account key
crypto_worker = AES.new(
    bytes.fromhex(account_key_secret),
    AES.MODE_CBC,
    initialization_vector
)
account_key = crypto_worker.decrypt(bytes.fromhex(encrypted_account_key)).hex()[0:64]
print("Decrypted account key", account_key)

# Generate secret key for mnemonic with a SHA3 hash
secret_key = keccak.new(digest_bits=256)
secret_key.update('{}encryption'.format(account_key).encode('utf-8'))
mnemonic_secret_key = secret_key.hexdigest()
print("mnemonicSecretKey", mnemonic_secret_key, len(mnemonic_secret_key))

# Use initialization vector and secret key to decrypt the mnemonic
crypto_worker = AES.new(
    bytes.fromhex(mnemonic_secret_key),
    AES.MODE_CBC,
    encrypted_mnemonic_iv
)
mnemonicHex = crypto_worker.decrypt(bytes.fromhex(encrypted_mnemonic)).hex()[0:296]
mnemonic = bytes.fromhex(mnemonicHex).decode('utf-8')
print("Decrypted mnemonic hex", mnemonicHex)
print("Decrypted mnemonic", mnemonic)

File: javascript/examples/decryptMnemonic.js
import axios from "axios";
import CryptoJS from "crypto-js";
import { apiToken, exchangePassword } from '../config.js';


async function decryptEncryptedBlob() {
  // Encrypted blob URL
  const url = "https://localcoinswap.com/api/v2/profile/encrypted-blob/";

  // Make a get request to blob URL for account data
  const result = await axios.get(url, {headers: {"Authorization": `Token ${apiToken}`}})

  // Parse encrypted parameters from GET request
  const saltHex = result.data.salt;
  const iterations = result.data.iterations;
  const accountKeyIVHex = result.data.initialization_vector;
  const encryptedAccountKey = result.data.encrypted_account_key;

  const encryptedMnemonicHex = result.data.encrypted_mnemonic;
  const mnemonicIVHex = result.data.encrypted_mnemonic_iv;

  // Combine salt and plaintext account password in PBKDF2 function
  // to create a secret key.
  const salt = CryptoJS.enc.Hex.parse(saltHex);
  const accountKeySecret = CryptoJS.PBKDF2(exchangePassword, salt, {
    keySize: 256 / 32,
    iterations,
  });
  console.log('accountKeySecret PBKDF2', CryptoJS.enc.Hex.stringify(accountKeySecret))

  // Use initialization vector and secret key to decrypt the account key
  const accountKeyParams = CryptoJS.lib.CipherParams.create({
    ciphertext: CryptoJS.enc.Hex.parse(encryptedAccountKey),
  });
  const accountKey = CryptoJS.AES.decrypt(accountKeyParams, accountKeySecret, {
    mode: CryptoJS.mode.CBC,
    iv: CryptoJS.enc.Hex.parse(accountKeyIVHex),
  });
  const accountKeyHexString = CryptoJS.enc.Hex.stringify(accountKey);
  console.log(`Decrypted account key ${accountKeyHexString}`);

  // Generate secret key for mnemonic with a SHA3 hash
  const mnemonicSecretKey = CryptoJS.SHA3(`${accountKeyHexString}encryption`, { outputLength: 256 });
  console.log(`mnemonicSecretKey ${mnemonicSecretKey}`);

  // Use initialization vector and secret key to decrypt the mnemonic
  const mnemonicParams = CryptoJS.lib.CipherParams.create({
    ciphertext: CryptoJS.enc.Hex.parse(encryptedMnemonicHex),
  });
  const mnemonic = CryptoJS.AES.decrypt(mnemonicParams, mnemonicSecretKey, {
    mode: CryptoJS.mode.CBC,
    iv: CryptoJS.enc.Hex.parse(mnemonicIVHex),
  });
  const mnemonicHex = CryptoJS.enc.Hex.stringify(mnemonic);
  const mnemonicString = CryptoJS.enc.Utf8.stringify(mnemonic);
  console.log(`Decrypted mnemonic hex ${mnemonicHex}`);
  console.log(`Decrypted mnemonic ${mnemonicString}`);
}

decryptEncryptedBlob();
